src/GNDStk/gnds-tree-node.hpp

```diff
@@ -0,0 +1,444 @@
+// Node (templated)
+// Then node = Node<>
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+class Node {
+   // for internal use
+   static bool static_found;
+
+   using pair = std::pair<std::string,std::string>;
+   using sptr = std::shared_ptr<Node>;
+
+public:
+
+   // Simple node for our tree structure:
+   //    name
+   //    metadata
+   //    children
+   std::string name;
+   MCON<pair,std::allocator<pair>> metadata;
+   CCON<sptr,std::allocator<sptr>> children;
+
+   // normalize
+   void normalize();
+
+   // push metadatum
+   pair &push(const std::string &key, const std::string &value)
+   {
+      metadata.push_back(std::make_pair(key,value));
+      return metadata.back();
+   }
+
+   pair &push(const pair &m)
+   {
+      metadata.push_back(m);
+      return metadata.back();
+   }
+
+   template<class T>
+   pair &push(const std::string &key, const T &value)
+   {
+      std::ostringstream oss;
+      oss << value;
+      return push(key,oss.str());
+   }
+
+   template<class T>
+   pair &push(const std::pair<std::string,T> &m)
+   {
+      std::ostringstream oss;
+      oss << m.second;
+      return push(pair(m.first,oss.str()));
+   }
+
+   // push child
+   Node &push(const std::string &name = "")
+   {
+      children.push_back(std::make_shared<Node<MCON,CCON>>());
+      children.back()->name = name;
+      return *children.back();
+   }
+
+   // write
+   bool write(const char * const file, const int level = 0) const;
+   bool write(const std::string &file, const int level = 0) const
+   {
+      return write(file.c_str(),level);
+   }
+   std::ostream &write(std::ostream &, const int level = 0) const;
+
+   // leaf?
+   bool leaf() const
+   {
+      return children.size() == 0;
+   }
+
+
+   // ------------------------
+   // meta()
+   // ------------------------
+
+   // for string
+   const std::string &meta(
+      const std::string &str,
+      bool &found = static_found // importantly, a reference; see below
+   ) const {
+      // search
+      for (auto &m : metadata)
+         if (m.first == str)
+            return found = true, m.second;
+
+      // not found
+      found = false;
+      static const std::string empty = "";
+
+      // If a "found" flag was NOT sent to this function, we'll emit an error.
+      // Otherwise, we'll assume that the caller intends to deal with the issue
+      // in its own way, and therefore that we shouldn't produce such an error.
+      // Note that the question of whether a "found" flag was sent is determined
+      // by looking at its address. This is entirely different from the question
+      // of what found's value should be. Its value is set either way.
+      if (&found == &static_found) {
+         // found references our private default - so, caller didn't send one
+         error(
```

Comment:
I completely agree; we're on the same general page with this. As I mentioned elsewhere, in a long comment, at the moment I've got simple assert()s where I know that we'll eventually want something more consistent with project policy (typically C++ exceptions). All along, those have been there as placeholders so as to avoid getting distracted with details rather than fleshing out an overall consistent look and feel to the API. I've often found that it's more efficient to wrote a boatload of real content, then sweep back through everything and do all the diagnostic/error/warning/etc. business in one fell swoop. Writing all the diagnostic messages at once can work wonders in terms of ensuring that the diagnostics themselves are written consistently - with consistent wording and such between them. For example, we *don't* want "Error: could not open file" in one place, and "Error: no such file" in another place. Inconsistent language confuses people; they wonder if there's some minor difference between this or that, a difference that might bite them in some other way. There's a method in my apparent madness for tossing in lots of placeholders (like the asserts for most errors right now), and focusing on code flow and design, as opposed to completely finishing a particular function, error messages and all, and then moving on to the next function. It's a somewhat nonlinear way of writing things, perhaps not a workflow you're accustomed to seeing, but I've found it to work well.